File: split/shownodelist.pas
procedure shownodelist(p: integer);
label
  10;
var
  n: integer;
  g: real;
begin
  if (poolptr - strstart[strptr]) > depththreshold then
  begin
    if p > 0 then
      print(314);
    goto 10;
  end;
  n := 0;
  while p > memmin do
  begin
    println;
    printcurrentstring;
    if p > memend then
    begin
      print(315);
      goto 10;
    end;
    n := n + 1;
    if n > breadthmax then
    begin
      print(316);
      goto 10;
    end;
    {183:}
if (p >= himemmin) then
      printfontandchar(p)
    else
      case mem[p].hh.b0 of
        0, 1, 13:{184:}

        begin
          if mem[p].hh.b0 = 0 then
            printesc(104)
          else if mem[p].hh.b0 = 1 then
            printesc(118)
          else
            printesc(318);
          print(319);
          printscaled(mem[p + 3].int);
          printchar(43);
          printscaled(mem[p + 2].int);
          print(320);
          printscaled(mem[p + 1].int);
          if mem[p].hh.b0 = 13 then{185:}

          begin
            if mem[p].hh.b1 <> 0 then
            begin
              print(
                286);
              printint(mem[p].hh.b1 + 1);
              print(322);
            end;
            if mem[p + 6].int <> 0 then
            begin
              print(323);
              printglue(mem[p + 6].int, mem[p + 5].hh.b1, 0);
            end;
            if mem[p + 4].int <> 0 then
            begin
              print(324);
              printglue(mem[p + 4].int, mem[p + 5].hh.b0, 0);
            end;
          end{:185}

          else
          begin
            {186:}
g := mem[p + 6].gr;
            if (g <> 0.0) and (mem[p + 5].hh.b0 <> 0) then
            begin
              print(325);
              if mem[p + 5].hh.b0 = 2 then
                print(326);
              if abs(mem[p + 6].int) < 1048576 then
                print(327)
              else if abs(g) > 20000.0 then
              begin
                if g > 0.0 then
                  printchar(62)
                else
                  print(328);
                printglue(20000 * 65536, mem[p + 5].hh.b1, 0);
              end
              else
                printglue(round(65536 * g), mem[p + 5].hh.b1, 0);
            end{:186}
;
            if mem[p + 4].int <> 0 then
            begin
              print(321);
              printscaled(mem[p + 4].int);
            end;
          end;
          begin
            begin
              strpool[poolptr] := 46;
              poolptr := poolptr + 1;
            end;
            shownodelist(mem[p + 5].hh.rh);
            poolptr := poolptr - 1;
          end;
        end{:184}
;
        2:{187:}

        begin
          printesc(329);
          printruledimen(mem[p + 3].int);
          printchar(43);
          printruledimen(mem[p + 2].int);
          print(320);
          printruledimen(mem[p + 1].int);
        end{:187}
;
        3:{188:}

        begin
          printesc(330);
          printint(mem[p].hh.b1 - 0);
          print(331);
          printscaled(mem[p + 3].int);
          print(332);
          printspec(mem[p + 4].hh.rh, 0);
          printchar(44);
          printscaled(mem[p + 2].int);
          print(333);
          printint(mem[p + 1].int);
          begin
            begin
              strpool[poolptr] := 46;
              poolptr := poolptr + 1;
            end;
            shownodelist(mem[p + 4].hh.lh);
            poolptr := poolptr - 1;
          end;
        end{:188}
;
        8:{1356:}
case mem[p].hh.b1 of
            0:
            begin
              printwritewhatsit(1284, p);
              printchar(61);
              printfilename(mem[p + 1].hh.rh, mem[p + 2].hh.lh, mem[p + 2].hh.rh);
            end;
            1:
            begin
              printwritewhatsit(594, p);
              printmark(mem[p + 1].hh.rh);
            end;
            2: printwritewhatsit(1285, p);
            3:
            begin
              printesc(1286);
              printmark(mem[p + 1].hh.rh);
            end;
            4:
            begin
              printesc(1288);
              printint(mem[p + 1].hh.rh);
              print(1291);
              printint(mem[p + 1].hh.b0);
              printchar(44);
              printint(mem[p + 1].hh.b1);
              printchar(41);
            end;
            others: print(1292)
          end{:1356}
;
        10:{189:}
if mem[p].hh.b1 >= 100 then{190:}

          begin
            printesc(338);
            if mem[p].hh.b1 = 101 then
              printchar(99)
            else if mem[p].hh.b1 = 102 then
              printchar(120);
            print(339);
            printspec(mem[p + 1].hh.lh, 0);
            begin
              begin
                strpool[poolptr] := 46;
                poolptr := poolptr + 1;
              end;
              shownodelist(mem[p + 1].hh.rh);
              poolptr := poolptr - 1;
            end;
          end{:190}

          else
          begin
            printesc(334);
            if mem[p].hh.b1 <> 0 then
            begin
              printchar(40);
              if mem[p].hh.b1 < 98 then
                printskipparam(mem[p].hh.b1 - 1)
              else if mem[p].hh.b1 = 98 then
                printesc(335)
              else
                printesc(336);
              printchar(41);
            end;
            if mem[p].hh.b1 <> 98 then
            begin
              printchar(32);
              if mem[p].hh.b1 < 98 then
                printspec(mem[p + 1].hh.lh, 0)
              else
                printspec(mem[p + 1].hh.lh, 337);
            end;
          end{:189}
;
        11:{191:}
if mem[p].hh.b1 <> 99 then
          begin
            printesc(340);
            if mem[p].hh.b1 <> 0 then
              printchar(32);
            printscaled(mem[p + 1].int);
            if mem[p].hh.b1 = 2 then
              print(341);
          end
          else
          begin
            printesc(342);
            printscaled(mem[p + 1].int);
            print(337);
          end{:191}
;
        9:{192:}

        begin
          printesc(343);
          if mem[p].hh.b1 = 0 then
            print(344)
          else
            print(345);
          if mem[p + 1].int <> 0 then
          begin
            print(346);
            printscaled(mem[p + 1].int);
          end;
        end{:192}
;
        6:{193:}

        begin
          printfontandchar(p + 1);
          print(347);
          if mem[p].hh.b1 > 1 then
            printchar(124);
          fontinshortdisplay := mem[p + 1].hh.b0;
          shortdisplay(mem[p + 1].hh.rh);
          if odd(mem[p].hh.b1) then
            printchar(124);
          printchar(41);
        end{:193}
;
        12:{194:}

        begin
          printesc(348);
          printint(mem[p + 1].int);
        end{:194}
;
        7:{195:}

        begin
          printesc(349);
          if mem[p].hh.b1 > 0 then
          begin
            print(350);
            printint(mem[p].hh.b1);
          end;
          begin
            begin
              strpool[poolptr] := 46;
              poolptr := poolptr + 1;
            end;
            shownodelist(mem[p + 1].hh.lh);
            poolptr := poolptr - 1;
          end;
          begin
            strpool[poolptr] := 124;
            poolptr := poolptr + 1;
          end;
          shownodelist(mem[p + 1].hh.rh);
          poolptr := poolptr - 1;
        end{:195}
;
        4:{196:}

        begin
          printesc(351);
          printmark(mem[p + 1].int);
        end{:196}
;
        5:{197:}

        begin
          printesc(352);
          begin
            begin
              strpool[poolptr] := 46;
              poolptr := poolptr + 1;
            end;
            shownodelist(mem[p + 1].int);
            poolptr := poolptr - 1;
          end;
        end{:197}
;
        {690:}
14: printstyle(mem[p].hh.b1);
        15:{695:}

        begin
          printesc(525);
          begin
            strpool[poolptr] := 68;
            poolptr := poolptr + 1;
          end;
          shownodelist(mem[p + 1].hh.lh);
          poolptr := poolptr - 1;
          begin
            strpool[poolptr] := 84;
            poolptr := poolptr + 1;
          end;
          shownodelist(mem[p + 1].hh.rh);
          poolptr := poolptr - 1;
          begin
            strpool[poolptr] := 83;
            poolptr := poolptr + 1;
          end;
          shownodelist(mem[p + 2].hh.lh);
          poolptr := poolptr - 1;
          begin
            strpool[poolptr] := 115;
            poolptr := poolptr + 1;
          end;
          shownodelist(mem[p + 2].hh.rh);
          poolptr := poolptr - 1;
        end{:695}
;
        16, 17, 18, 19, 20, 21, 22, 23, 24, 27, 26, 29, 28, 30, 31:{696:}

        begin
          case mem[p].hh.b0 of
            16: printesc(865);
            17: printesc(866);
            18: printesc(867);
            19: printesc(868);
            20: printesc(869);
            21: printesc(870);
            22: printesc(871);
            23: printesc(872);
            27: printesc(873);
            26: printesc(874);
            29: printesc(539);
            24:
            begin
              printesc(533);
              printdelimiter(p + 4);
            end;
            28:
            begin
              printesc(508);
              printfamandchar(p + 4);
            end;
            30:
            begin
              printesc(875);
              printdelimiter(p + 1);
            end;
            31:
            begin
              printesc(876);
              printdelimiter(p + 1);
            end;
          end;
          if mem[p].hh.b1 <> 0 then
            if mem[p].hh.b1 = 1 then
              printesc(877)
            else
              printesc(878);
          if mem[p].hh.b0 < 30 then
            printsubsidiarydata(p + 1, 46);
          printsubsidiarydata(p + 2, 94);
          printsubsidiarydata(p + 3, 95);
        end{:696}
;
        25:{697:}

        begin
          printesc(879);
          if mem[p + 1].int = 1073741824 then
            print(880)
          else
            printscaled(mem[p + 1].int);
          if (mem[p + 4].qqqq.b0 <> 0) or (mem[p + 4].qqqq.b1 <> 0) or (mem[p + 4].qqqq.b2 <> 0) or (mem[p + 4].qqqq.b3 <> 0) then
          begin
            print(881);
            printdelimiter(p + 4);
          end;
          if (mem[p + 5].qqqq.b0 <> 0) or (mem[p + 5].qqqq.b1 <> 0) or (mem[p + 5].qqqq.b2 <> 0) or (mem[p + 5].qqqq.b3 <> 0) then
          begin
            print(882);
            printdelimiter(p + 5);
          end;
          printsubsidiarydata(p + 2, 92);
          printsubsidiarydata(p + 3, 47);
        end{:697}
;
        {:690}
others: print(317)
      end{:183}
;
    p := mem[p].hh.rh;
  end;
  10: ;
end;

File: cpp/shownodelist.cpp
#include "shownodelist.h"
#include "print.h"
#include "println.h"
#include "printesc.h"
#include "printchar.h"
#include "printfontandchar.h"
#include "printcurrentstring.h"
#include "printscaled.h"
#include "printglue.h"
#include "printint.h"
#include "printspec.h"
#include "printwritewhatsit.h"
#include "printdelimiter.h"
#include "printsubsidiarydata.h"
#include "printstyle.h"
#include "printmark.h"
#include "shortdisplay.h"
#include "printskipparam.h"
#include "printruledimen.h"
#include "printfilename.h"
#include "printfamandchar.h"
#include <cmath>

void shownodelist(int p)
{
	if (poolptr - strstart[strptr] > depththreshold)
	{
		if (p > 0)
			print(" []");
		return;
	}
	int n = 0;
	while (p > memmin)
	{
		println();
		printcurrentstring();
		if (p > memend)
		{
			print("Bad link, display aborted.");
			return;
		}
		n++;
		if (n > breadthmax)
		{
			print("etc.");
			return;
		}
		if (p >= himemmin)
			printfontandchar(p);
		else
			switch (type(p))
			{
				case hlist_node:
				case vlist_node:
				case unset_node:
					if (type(p) == hlist_node)
						printesc("h");
					else 
						if (type(p) == vlist_node)
							printesc("v");
						else
							printesc("unset");
					print("box(");
					printscaled(height(p));
					printchar('+');
					printscaled(depth(p));
					print(")x");
					printscaled(width(p));
					if (type(p) == unset_node)
					{
						if (subtype(p))
						{
							print(" (");
							printint(subtype(p)+1);
							print(" columns)");
						}
						if (glue_stretch(p))
						{
							print(", stretch ");
							printglue(glue_stretch(p), glue_order(p), 0);
						}
						if (glue_shrink(p))
						{
							print(", shrink ");
							printglue(glue_shrink(p), glue_sign(p), 0);
						}
					}
					else
					{
						g = glue_set(p);
						if (g && glue_sign(p))
						{
							print(", glue set ");
							if (glue_sign(p) == shrinking)
								print("- ");
							if (!std::isfinite(glue_set(p)))
								print("?.?");
							else 
								if (abs(g) > 20000.0)
								{
									if (g > 0.0)
										printchar('>');
									else
										print("< -");
									printglue(20000*0x1'00'00, glue_order(p), 0);
								}
								else
									printglue(round(0x1'00'00*g), glue_order(p), 0);
						}
						if (shift_amount(p))
						{
							print(", shifted ");
							printscaled(shift_amount(p));
						}
					}
					strpool[poolptr++] = '.';
					shownodelist(list_ptr(p));
					poolptr--;
					break;
				case rule_node:
					printesc("rule(");
					printruledimen(width(p));
					printchar('+');
					printruledimen(depth(p));
					print(")x");
					printruledimen(width(p));
					break;
				case ins_node:
					printesc("insert");
					printint(subtype(p));
					print(", natural size ");
					printscaled(height(p));
					print("; split(");
					printspec(split_top_ptr(p), 0);
					printchar(',');
					printscaled(depth(p));
					print("); float cost ");
					printint(float_cost(p));
					strpool[poolptr++] = '.';
					shownodelist(ins_ptr(p));
					poolptr--;
					break;
				case whatsit_node:
					switch (subtype(p))
					{
						case 0:
							printwritewhatsit("openout", p);
							printchar('=');
							printfilename(link(p+1), info(p+2), link(p+2));
							break;
						case 1:
							printwritewhatsit("write", p);
							printmark(link(p+1));
							break;
						case 2: 
							printwritewhatsit("closeout", p); 
							break;
						case 3:
							printesc("special");
							printmark(link(p+1));
							break;
						case 4:
							printesc("setlanguage");
							printint(link(p+1));
							print(" (hyphenmin"); 
							printint(type(p+1));
							printchar(',');
							printint(subtype(p+1));
							printchar(')');
							break;
						default: 
							print("whatsit?");
					}
					break;
				case glue_node:
					if (subtype(p) >= 100)
					{
						printesc("");
						if (subtype(p) == 101)
							printchar('c');
						else 
							if (subtype(p) == 102)
								printchar('x');
						print("leaders ");
						printspec(info(p+1), 0);
						strpool[poolptr++] = '.';
						shownodelist(link(p+1));
						poolptr--;
					}
					else
					{
						printesc("glue");
						if (subtype(p))
						{
							printchar('(');
							if (subtype(p) < 98)
								printskipparam(subtype(p)-1);
							else 
								if (subtype(p) == 98)
								printesc("nonscript");
							else
								printesc("mskip");
							printchar(')');
						}
						if (subtype(p) != 98)
						{
							printchar(' ');
							if (subtype(p) < 98)
								printspec(info(p+1), 0);
							else
								printspec(info(p+1), "mu");
						}
					}
				case kern_node:
					if (subtype(p) != 99)
					{
						printesc("kern");
						if (subtype(p))
							printchar(' ');
						printscaled(width(p));
						if (subtype(p) == 2)
							print(" (for accent)");
					}
					else
					{
						printesc("mkern");
						printscaled(width(p));
						print("mu");
					}
					break;
				case math_node:
					printesc("math");
					if (subtype(p) == 0)
						print("on");
					else
						print("off");
					if (width(p))
					{
						print(", surrounded ");
						printscaled(width(p));
					}
					break;
				case ligature_node:
					printfontandchar(p+1);
					print(" (ligature ");
					if (subtype(p) > 1)
						printchar('|');
					fontinshortdisplay = type(p+1);
					shortdisplay(link(p+1));
					if (subtype(p)%2)
						printchar('|');
					printchar(')');
					break;
				case penalty_node:
					printesc("penalty ");
					printint(penalty(p));
					break;
				case disc_node:
					printesc("discretionary");
					if (subtype(p) > 0)
					{
						print(" replacing ");
						printint(subtype(p));
					}
					strpool[poolptr++] = '.';
					shownodelist(info(p+1));
					poolptr--;
					strpool[poolptr++] = '|';
					shownodelist(link(p+1));
					poolptr--;
					break;
				case mark_node:
					printesc("mark");
					printmark(mark_ptr(p));
					break;
				case adjust_node:
					printesc("vadjust");
					strpool[poolptr++] = '.';
					shownodelist(adjust_ptr(p));
					poolptr--;
					break;
				case style_node: 
					printstyle(subtype(p));
					break;
				case choice_node:
					printesc("mathchoice");
					strpool[poolptr++] = 'D';
					shownodelist(info(p+1));
					poolptr--;
					strpool[poolptr++] = 'T';
					shownodelist(link(p+1));
					poolptr--;
					strpool[poolptr++] = 'S';
					shownodelist(info(p+2));
					poolptr--;
					strpool[poolptr++] = 's';
					shownodelist(link(p+2));
					poolptr--;
					break;
				case ord_noad:
				case op_noad:
				case bin_noad:
				case rel_noad:
				case open_noad:
				case close_noad:
				case punct_noad:
				case inner_noad:
				case radical_noad:
				case over_noad:
				case under_noad:
				case vcenter_noad:
				case accent_noad:
				case left_noad:
				case right_noad:
					switch (type(p))
					{
						case ord_noad: 
							printesc("mathord");
							break;
						case op_noad: 
							printesc("mathop");
							break;
						case bin_noad: 
							printesc("mathbin");
							break;
						case rel_noad: 
							printesc("mathrel");
							break;
						case open_noad: 
							printesc("mathopen");
							break;
						case close_noad: 
							printesc("mathclose");
							break;
						case punct_noad: 
							printesc("mathpunct");
							break;
						case inner_noad: 
							printesc("mathinner");
							break;
						case over_noad: 
							printesc("overline");
							break;
						case under_noad: 
							printesc("underline");
							break;
						case vcenter_noad: 
							printesc("vcenter");
							break;
						case radical_noad:
							printesc("radical");
							printdelimiter(left_delimiter(4));
							break;
						case accent_noad:
							printesc("accent");
							printfamandchar(accent_chr(p));
							break;
						case left_noad:
							printesc("left");
							printdelimiter(delimiter(p));
							break;
						case right_noad:
							printesc("right");
							printdelimiter(delimiter(p));
					}
					if (subtype(p))
						if (subtype(p) == limits)
							printesc("limits");
						else
							printesc("nolimits");
					if (type(p) < left_noad)
						printsubsidiarydata(nucleus(p), '.');
					printsubsidiarydata(supscr(p), '^');
					printsubsidiarydata(subscr(p), '_');
					break;
				case fraction_noad:
					printesc("fraction, thickness ");
					if (new_hlist(p) == default_code)
						print("= default");
					else
						printscaled(new_hlist(p));
					if (small_fam(left_delimiter(p))
					 || small_char(left_delimiter(p))
					 || large_fam(left_delimiter(p))
					 || large_char(left_delimiter(p)))
					{
						print(", left-delimiter ");
						printdelimiter(left_delimiter(p));
					}
					if (small_fam(right_delimiter(p))
					 || small_char(right_delimiter(p))
					 || large_fam(right_delimiter(p))
					 || large_char(right_delimiter(p)))
					{
						print(", right-delimiter ");
						printdelimiter(right_delimiter(p));
					}
					printsubsidiarydata(numerator(p), '\\');
					printsubsidiarydata(denominator(p), '/');
					break;
				default: 
					print("Unknown node type!");
			}
		p = link(p);
	}
}
